Re-adding the trivia command following the new architecture + UT

jjaxu opened the pull request:
Usage:
- `/trivia help` to see the help message associated with the trivia interface
- `/trivia [ <category> | any ] [ <difficulty> | any ]` to fetch a random question from a specified category and difficulty, or `any` for a randomized flag. Available categories and difficulty are displayed from the trivia help command. If omitted, it will default to "any"

jjaxu committed:
Re-adding the trivia command following the new architecture + UT

george-lim commented on commands/command_parser.py:
```diff
@@ -20,9 +20,13 @@ class CommandParser:
     @classmethod
     def parse_command(cls, bot_query: BotQuery) -> BotCommand:
+        if bot_query.has_callback_query:
+            cmd_name = bot_query.callback_query.callback_data.rsplit('/')[-1].lower()
```
Can you either make `.rsplit('/')[-1].lower()` into a separate function or comment what this is doing?

Something like:
```
# Returns the case insensitive command name from the query callback data. Eg. `/command` -> `command`
```

george-lim commented on commands/command_parser.py:
```diff
@@ -20,9 +20,13 @@ class CommandParser:
     def parse_command(cls, bot_query: BotQuery) -> BotCommand:
+        if bot_query.has_callback_query:
+            cmd_name = bot_query.callback_query.callback_data.rsplit('/')[-1].lower()
+            return AVAILABLE_COMMANDS[cmd_name](bot_query) if cmd_name in AVAILABLE_COMMANDS else InvalidCommand(bot_query)
```
Nitpick: I'm not personally a fan of this ternary statement because it is way too dense. You can decide if you want to keep this the way it is, but I would have done:
```
if cmd_name:
  return AVAILABLE_COMMANDS[cmd_name](bot_query)
return InvalidCommand(bot_query)
```

Here, control flow is shown through the flow of the statements, instead of being paired with the actual return logic. I generally reserve ternary statements for completely trivial return values (eg. return 1 if is_seen else 0)

george-lim commented on commands/command_parser.py:
```diff
@@ -20,9 +20,13 @@ class CommandParser:
         args = str(bot_query.message).strip().split()
         cmd_name = args[0].lower() if len(args) > 0 else ""
 
         if cmd_name.startswith("/") and cmd_name[1:] in AVAILABLE_COMMANDS:
```
can we make `cmd_name[1:]` a variable please?

george-lim commented on commands/trivia_command.py:
```diff
@@ -1,15 +1,88 @@
+
     def execute(self):
-        pass
+        if self.query.has_callback_query:
+            self.handle_callback_query()
+            return
```
This looks like it should just be a separate method. How about having two methods: `execute` and `execute_callback_query`. Main will automatically call `execute_callback_query` on any command that supports callback_queries, when the parser determines that the query is a callback_query. I don't like funnelling everything into execute, but then doing an early return and calling another method.

george-lim commented on commands/trivia_command.py:
```diff
@@ -1,15 +1,88 @@
+        if self.query.has_callback_query:
+            self.handle_callback_query()
+            return
+        category, difficulty = None, None
```
Nitpick: I didn't know you could do this, but it also looks very weird to me. It potentially looks like you are assigning both values as a pair of `None`. I always prefer keeping declarations on separate lines, but if you prefer this that's fine.

george-lim commented on commands/trivia_command.py:
```diff
@@ -1,15 +1,88 @@
+        if not self.trivia:
+            return
+
+        correct_ans = chr(self.trivia.correct_answer_index() + 65)
+        buttons = [ {"text": chr(i+65), "callback_data": f"{chr(i+65)}{correct_ans}/trivia" } for i in range(4) ]
```
too much logic here that isn't clear. Please either comment what this is doing, or break these down into variables. I know that you are using casting to convert the index into a letter, but I think this can be turned into a function `answer_letter(self.trivia)` to make it clear. Also, I know what the second line is doing, but you might want to at least explain what `f"{chr(i+65)}{correct_ans}/trivia" } for i in range(4)` is doing. It is way too complicated.

george-lim commented on trivia.py:
```diff
@@ -3,49 +3,58 @@
+
+    @classmethod
+    def get_help_text(cls) -> str:
+        category_list = [("\* " + item["name"]) for item in (cls.get_category_dict().get("trivia_categories", []))]
```
Too complicated. Please break this down or explain what it is doing.

george-lim commented on trivia.py:
```diff
@@ -3,49 +3,58 @@
         difficulty_list = [("\* " + item.capitalize()) for item in cls.difficulty_list]
-        return "Usage: */trivia [category] [difficulty]*\n\n*Categories:*\n{}\n\n*Difficulties:*\n{}".format("\n".join(category_list), "\n".join(difficulty_list))
+        nl = '\n'
+        return f"Usage: */trivia [category] [difficulty]*\n\n*Categories:*\n{nl.join(category_list)}\n\n*Difficulties:*\n{nl.join(difficulty_list)}"
```
please turn this into a method, this is way too difficult to read.

george-lim reviewed (changes requested):
I think you have a slight tendency to use SUPER condensed code at times without much explanation. Remember that while you may be able to read and understand what is going on, another developer reviewing your code might not be. To combat this, either comment on complex code segments in your codebase, or refactor / abstract them into methods and break lines down using intermediary variables.

How about instead of calling `.lower()` everywhere in each command, you automatically call `.lower()` on relevant parts (like args, command name, etc.) in your parser? Simplifies the amount of `.lower()` in your codebase and honestly, it's a formatting issue that your parser should take care of. Any additional parsing / input formatting can be handled by the command if it is super specific.

jjaxu committed:
Tweaking code according to review comments
- re-factored `TriviaCommand`'s `execute()` logic to be cleaner
- Adding some documentation and extra local variables to make condense parts more readable

george-lim reviewed (approved):
Nice 👍
